cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -276,24 +338,32 @@ private long processResponse(Map<String, List<MonitorMessage>> monitorResponses)
         }
       }
 
-      String topic = requestKeyToLocalTopic.get(topicConfig);
-      latestPublishTime = Math.max(publish(topicConfig, topic, monitorMessages), latestPublishTime);
+      Long publishTime = callWithRetries((Retries.Callable<Long, TransactionFailureException>) () -> {
+        Transactionals.execute(transactional, context -> {
+          return publish(topicConfig, requestKeyToLocalTopic.get(topicConfig), monitorMessages, context);
+        });
+        return null;
+      });
+
+      latestPublishTime = publishTime == null ? latestPublishTime : Math.max(publishTime, latestPublishTime);
     }
+
     return latestPublishTime;
   }
 
   /**
-   * Publish the given set of messages to the local TMS.
+   * Transactionally publish the given set of messages to the local TMS.
    *
    * @return the latest message publish time or {@code -1L} if there is no message to process
    */
-  private long publish(String topicConfig, String topic, List<MonitorMessage> messages) throws Exception {
-    // TODO publish messages transactionally along with offset table updates
+  private long publish(String topicConfig, String topic,
+                       List<MonitorMessage> messages, DatasetContext context) throws Exception {
+    MessagePublisher messagePublisher = messagingContext.getMessagePublisher();
     messagePublisher.publish(NamespaceId.SYSTEM.getNamespace(), topic,
                              messages.stream().map(s -> s.getMessage().getBytes(StandardCharsets.UTF_8)).iterator());
-
     MonitorMessage lastMessage = updateTopicToRequest(topicConfig, messages);
-
+    AppMetadataStore.create(cConf, context, dsFramework).persistSubscriberState(topicConfig, programRunId.getRun(),
```

Comment:
The `AppMetadataStore.create` is effective getting called inside a loop. Please refactor accordingly.